lib/controller/cvx.py

```diff
@@ -0,0 +1,173 @@
+import cvxpy as cp
+import autograd as AG
+import autograd.numpy as np 
+
+# mc = 0.2 # mass of the cart (kg)
+# mp = 0.1 # mass of the pole (kg)
+# l = 0.5 # distance to the center of mass (meters)
+# g = 9.81
+
+# xdot = dynamics(x, u)
+# x = [pos*3, vel*3, ang*3, angvel*3]
+# u = [gimbalx, gimbaly, rollthruster, thruster]
+def rot_matrix(rx,ry,rz):
+    return np.array([
+        [np.cos(rz)*np.cos(ry), np.cos(rz)*np.sin(ry)*np.sin(rx) - np.sin(rz)*np.cos(rx), np.cos(rz)*np.sin(ry)*np.cos(rx) + np.sin(rz)*np.sin(rx)],
+        [np.sin(rz)*np.cos(ry), np.sin(rz)*np.sin(ry)*np.sin(rx) + np.cos(rz)*np.cos(rx), np.sin(rz)*np.sin(ry)*np.cos(rx) - np.cos(rz)*np.sin(rx)],
+        [-np.sin(ry), np.cos(ry)*np.sin(rx), np.cos(ry)*np.cos(rx)]
+    ])
+
+def dynamics(X, U):
+    r = 0.35
+    M = 10
+    I = 1/2 * (M / 2) * r**2
+
+    x,y,z = X[0:3]
+    vx,vy,vz = X[3:6]
+    rx,ry,rz = X[6:9]
+    drx,dry,drz = X[9:12]
+    gimbalx, gimbaly, rollT, fT = U
+
+    # xyz are in inertial ref frame
+    # rotation are in body ref frame
+    # xyz pos has no effect on vel or angvel
+    # vx, vy, vz directly affect xyz
+    # acceleration depends on thrust + direction
+    # Rot(Body) @ (Rot(Gimbal) @ [0, 0, fT].T) - [0, 0, 9.81]
+    # angvel directly dependent on drx, dry, drz
+    # Angular acceleration dependent on torque from gimbaled motor
+    # tau = rF sin theta -> angaccel = rF sin theta / I, I is rotational inertia
+    # We assume rocket body is solid cylinder and legs are negligible (change this for real world!)
+    accel = rot_matrix(rx,ry,rz) @ (rot_matrix(gimbalx, gimbaly, 0) @ np.array([0, 0, fT])) / M - np.array([0, 0, 9.81])
+    arx = r * fT * np.sin(gimbalx) / I
+    ary = r * fT * np.sin(gimbaly) / I
+    arz = rollT
+    print(f"arz {arz}, drz {drz}")
+    return np.array([
+        vx, vy, vz,
+        accel[0], accel[1], accel[2],
+        drx, dry, drz,
+        arx, ary, arz
+    ])
+
+def dynamics_rk4(x, u, dt):
+    f1 = dt*dynamics(x, u)
+    f2 = dt*dynamics(x + f1/2, u)
+    f3 = dt*dynamics(x + f2/2, u)
+    f4 = dt*dynamics(x + f3, u)
+    return x + (1/6)*(f1 + 2*f2 + 2*f3 + f4)
+
+# Neutral position to linearize around
+xgoal = np.zeros(12) # On ground, straight up, no movement
```

Comment:
Does this controller hard-code a "trajectory"?

Reply:
No, this specific controller targets a specific position to stabilize around. Trajectory tracking probably should have a slightly different formulation